File: ios/AIChallenge/Domain/Models/ContextStrategy.swift
import Foundation

enum ContextStrategy: String, Codable, CaseIterable, Identifiable, Sendable {
    case slidingWindow = "sliding_window"
    case stickyFacts = "sticky_facts"
    case branching

    var id: String { rawValue }

    var title: String {
        switch self {
        case .slidingWindow: "Window"
        case .stickyFacts: "Facts"
        case .branching: "Ветки"
        }
    }

    var fullTitle: String {
        switch self {
        case .slidingWindow: "Sliding Window"
        case .stickyFacts: "Sticky Facts"
        case .branching: "Branching"
        }
    }

    var explanation: String {
        switch self {
        case .slidingWindow: "Только 6 последних сообщений. Дёшево, но ранние решения исчезают безвозвратно."
        case .stickyFacts: "Важные данные живут в key-value памяти, рядом остаются 6 свежих сообщений."
        case .branching: "Checkpoint клонирует общий контекст в две независимые версии: MVP и Growth."
        }
    }
}

struct MemoryFact: Codable, Equatable, Identifiable, Sendable {
    let key: String
    let value: String
    var id: String { key }
}

struct StrategyBranch: Codable, Equatable, Identifiable, Sendable {
    let id: String
    let title: String
    let messageCount: Int
    let checkpointMessages: Int
}

struct ContextStrategyState: Codable, Equatable, Sendable {
    let sessionId: String
    let strategy: ContextStrategy
    let activeBranchId: String?
    let windowSize: Int
    let messages: [AIAgentMessage]
    let facts: [MemoryFact]
    let branches: [StrategyBranch]
    let lastUsage: ModelUsage
}

struct ContextStrategyExchange: Codable, Equatable, Sendable {
    let strategy: ContextStrategy
    let branchId: String?
    let exchange: AIAgentExchange
    let state: ContextStrategyState
}

struct StrategyBranchAnswer: Codable, Equatable, Identifiable, Sendable {
    let branchId: String
    let title: String
    let answer: String
    let usage: ModelUsage
    var id: String { branchId }
}

struct ContextStrategyResult: Codable, Equatable, Identifiable, Sendable {
    let strategy: ContextStrategy
    let answer: String
    let usage: ModelUsage
    let messagesKept: Int
    let factsKept: Int
    let branches: [StrategyBranchAnswer]
    let behavior: String
    var id: String { strategy.id }
}

struct ContextStrategyScore: Codable, Equatable, Identifiable, Sendable {
    let strategy: ContextStrategy
    let quality: Int
    let stability: Int
    let tokenEfficiency: Int
    let usability: Int
    let feedback: String
    var id: String { strategy.id }
}

struct ContextStrategyReview: Codable, Equatable, Sendable {
    let winner: ContextStrategy
    let verdict: String
    let differences: [String]
    let scores: [ContextStrategyScore]
    let recommendations: [String]
    let model: String
    let usage: ModelUsage
}

struct ContextStrategyComparison: Codable, Equatable, Sendable {
    let sessionId: String
    let scenario: [String]
    let question: String
    let results: [ContextStrategyResult]
    let review: ContextStrategyReview
}

File: backend/internal/domain/context_strategy.go
package domain

type ContextStrategy string

const (
	ContextStrategySlidingWindow ContextStrategy = "sliding_window"
	ContextStrategyStickyFacts   ContextStrategy = "sticky_facts"
	ContextStrategyBranching     ContextStrategy = "branching"
)

func (s ContextStrategy) IsValid() bool {
	return s == ContextStrategySlidingWindow || s == ContextStrategyStickyFacts || s == ContextStrategyBranching
}

type MemoryFact struct {
	Key   string `json:"key"`
	Value string `json:"value"`
}

type StrategyBranch struct {
	ID                 string `json:"id"`
	Title              string `json:"title"`
	MessageCount       int    `json:"messageCount"`
	CheckpointMessages int    `json:"checkpointMessages"`
}

type ContextStrategyState struct {
	SessionID      string           `json:"sessionId"`
	Strategy       ContextStrategy  `json:"strategy"`
	ActiveBranchID string           `json:"activeBranchId,omitempty"`
	WindowSize     int              `json:"windowSize"`
	Messages       []AgentMessage   `json:"messages"`
	Facts          []MemoryFact     `json:"facts"`
	Branches       []StrategyBranch `json:"branches"`
	LastUsage      Usage            `json:"lastUsage"`
}

type ContextStrategyExchange struct {
	Strategy ContextStrategy      `json:"strategy"`
	BranchID string               `json:"branchId,omitempty"`
	Exchange AgentExchange        `json:"exchange"`
	State    ContextStrategyState `json:"state"`
}

type StrategyBranchAnswer struct {
	BranchID string `json:"branchId"`
	Title    string `json:"title"`
	Answer   string `json:"answer"`
	Usage    Usage  `json:"usage"`
}

type ContextStrategyResult struct {
	Strategy     ContextStrategy        `json:"strategy"`
	Answer       string                 `json:"answer"`
	Usage        Usage                  `json:"usage"`
	MessagesKept int                    `json:"messagesKept"`
	FactsKept    int                    `json:"factsKept"`
	Branches     []StrategyBranchAnswer `json:"branches"`
	Behavior     string                 `json:"behavior"`
}

type ContextStrategyScore struct {
	Strategy        ContextStrategy `json:"strategy"`
	Quality         int             `json:"quality"`
	Stability       int             `json:"stability"`
	TokenEfficiency int             `json:"tokenEfficiency"`
	Usability       int             `json:"usability"`
	Feedback        string          `json:"feedback"`
}

type ContextStrategyReview struct {
	Winner          ContextStrategy        `json:"winner"`
	Verdict         string                 `json:"verdict"`
	Differences     []string               `json:"differences"`
	Scores          []ContextStrategyScore `json:"scores"`
	Recommendations []string               `json:"recommendations"`
	Model           string                 `json:"model"`
	Usage           Usage                  `json:"usage"`
}

type ContextStrategyComparison struct {
	SessionID  string                  `json:"sessionId"`
	WindowSize int                     `json:"windowSize"`
	Scenario   []string                `json:"scenario"`
	Question   string                  `json:"question"`
	Results    []ContextStrategyResult `json:"results"`
	Review     ContextStrategyReview   `json:"review"`
}
